### ISSUE TITLE
Relaxing target function requirements

### ISSUE THREAD

**Issue by guyio** [googleprojectzero/winafl#117]
Hey Ivan,

I think that many real world targets are managing some sort of global context which without process forking like AFL does it's quite hard or maybe even impossible to find a compliant target function for the winafl approach.

With that in mind, the good news is that many target applications can actually loop by themselves, for example a javascript loop inside a browser, a udp server that read packets and handles them.
If this is the case, the execution redirection can be avoided.

So my idea is to add another fuzzing mode 'in_app_persistency' which works exactly on these target traits,
The main idea is that each execution loop will do 2 things
1. read the target file (we don't care from which context)
2. enter the actual target function to be fuzzed (which won't require actual file reading inside)

Now, since the target program is going to loop by itself we need to account for 2 scenarios:
1. file was read and target func executed .
2. file was read and target func wasn't executed.

It's possible to account for both with adding some code to the 'createfile' hook inside the winafl client to notify the fuzz server we reached it, and get a message from the fuzz server to continue.

I've forked your project and added this functionality under a new fuzzing mode and it works great for adobe reader (for which i couldn't find a compliant target function to the traditional approach).
Also I've 'fixed' a corner case in the run_target function where it's possible that between the child process liveness check and the ReadFile from pipe the child dies and the fuzz server is stuck.

If you think it will be helpful enough for more researchers I can beautify it a bit more and issue a pull request.

**Comment by ivanfratric** [comment]
Hey Guy,

This is interesting and I think the main challenge is doing this in a way that works (without having to make custom changes) for a lot of applications. As I'm not sure I fully understand all the details of your approach, could you maybe write more detailed steps for the fuzzing flow for the example targets you mentioned (browser, a udp server, adobe reader). Some things things that aren't clear:

 - In the scenarios you mentioned, how would file reading work, for example an udp server wouldn't read files typically and I'm not sure how you'd get e.g. Reader to loop over reading a file. I realize that data can be brought to target without file reading requirement with custom changes to WinAFL, but I'm not sure what a good application-agnostic way to do it would be.

- In what case would the file be read but target function not executed?

BTW I think I did something similar to what you described when I experimented with coverage-guided DOM fuzzing. My approach was to write a loop in JavaScript that creates an iframe with a sample and wait until it is fully loaded, then destroy the iframe and load a new one etc. Instead of looping with WinAfl, I just added a Math.acos() call in the JavaScript loop and I instrumented Math.acos() so that When Math.acos is reached, WinAFL starts collecting coverage. When it's reached again, this means the sample has finished executing, so report the coverage and start collecting it again for the new sample. However, this all seemed very web-browser specific so I didn't merge any of those changes with WinAFL.

Regarding a run_target race condition, could you submit a PR for the fix?

**Comment by ivanfratric** [comment]
Disregarding the "file be read but target function not executed" problem for a minute, do you think sufficient change to support such targets would be to introduce a flag to remove looping, that is skip https://github.com/ivanfratric/winafl/blob/master/winafl.c#L490 an the 2 lines before? This would change the meaning of a target function from "read a file, measure coverage and loop" to simply "measure coverage during this function". If this wouldn't be sufficient, why not? (I'm trying to figure out what the minimum set of changes required to support this is. And also if I understood what you are proposing clearly).

For the "file be read but target function not executed", I'm wondering if, instead of instrumenting createfile, as you suggested, it would make sense to specify another function in addition to target_method that would tell WinAfl, "target is ready to accept the next sample". This could be a function that reads a file down the line, but not always necessary. Reasoning
 - Sometimes a target might open the same file multiple times in a single loop
 - Might be easier to support targets that don't read files in the same process such as for fuzzing network apps

**Comment by guyio** [comment]
I've tried exactly what you just described. If your don't instrument the "read_input" function and just let the program loop, you will miss the input reading due to a race.
Take for example this scenatio:
1. afl_fuzz sets first input
2. afl_fuzz calls run_target
3. target app gets blocked on entry to target_function (pre_fuzz_handler)
4. afl_fuzz send a signal so that the pre_fuzz_handler continue exection (into the target func)
5. target app exits target func and sends a message over pipe to server
...here 2 things happen in parallel
6.a. server write file and wait for the pre_fuzz_handler message
6.b. target app reads file again and continue into pre_fuzz_handler

thus we must instrument the "read_input" function.

So basically I think the best solution to make it "generic" is to have 2 functions as you suggested
1. read_input
2. target_func

If we assume that the target app will only read input once and target_func will be executed with every cycle then there is no need for any change in the server. it could be easily done inside winafl.c alone by moving the read from pipe from the pre_fuzz_handler to the pre_read_input_handler.

maybe I can add an optional switch to let the read_input to be executed more than once and also support the "no target function reached in cycle" scenario?
I really like this approach as it lets users of winafl to harness many new targets out of the box.

**Comment by guyio** [comment]
True, It's possible to just switch "post_fuzz_handler" and "pre_fuzz_handler" to a single pre_fuzz_handler hook on the read_input function instead that sends "K" and waits for "F".
It will only require the server to remove the first iterations "K" from the pipe (easily achievable with PeekNamedPipe and making sure there are no bytes ready).

As of my understanding (which i hope is wrong), dynamorio's current instrumentation will only work for a single thread.
If this is correct, then it means that this solution will only work if read_input and target_func are on the same thread.
Unfortunately this isn't always the case.

If I'm wrong about this dr's assumption, then great, the problem is solved quite easily.

But anyhow, if the read_input might be called several times on each cycle it will be desirable to know when the target function is called so the server can know for sure a cycle has occured.

**Comment by ivanfratric** [comment]
Hey Guy, sorry about the delay. Actually, dynamorio's instrumentation will apply to all threads. There is actually a flag in WinAFL, -thread_coverage, that enables you (via thread local storage) to collect the coverage only from the thread that hit target function, but that is off by default.

"But anyhow, if the read_input might be called several times on each cycle it will be desirable to know when the target function is called so the server can know for sure a cycle has occured."

Yes, but earlier we discussed the problem if read_input gets called but target function is not called. I don't think we can reasonably cover all corner cases, and for simplicity, we should assume read_input gets called only once. Actually, with the approach discussed currently, read_input does not need to be a function that is related to input reading, it can be any function that is guaranteed to be called only once per loop iteration and I think such a function would be findable in most targets. Perhaps for clarity we should call it differently, loop_start, iteration_start or something similar. (perhaps loop_end/iteration_end flag can be offered as an alternative, the only difference would be that first 'K' signal).

**Comment by guyio** [comment]
Great then, I think this approach is very reasonable, I like the loop_start/loop_end concept.
I'll play with it a bit and issue a pr as soon as i have something ready.

**Comment by ifratric** [comment]
I believe this is resolved.
